### DISCUSSION TITLE
Artifacts on dielectric

### DISCUSSION THREAD

**Discussion by maurocolella** [RayTracing/raytracing.github.io#513] [Debugging]
Hi, thank you for this wonderful book series. It's Sunday. I have done what the title says: in one weekend, ported the code of book 1 to Python, as an exercise in both Python and raytracing.

I believe I made a mostly (very) straight port of everything including the dielectric material, but when transposing the exact code currently found in the book, now in comment at:
https://github.com/maurocolella/python-raytracer/blob/master/threed/dielectric_material.py

I get an awful artifact (see attached). 

I am not quite sure what is happening. I hastily reverted this to a version that produced the desired result from a previous iteration of the book, essentially based on implementations by other readers.

Help with this would be appreciated greatly.
![render](https://user-images.githubusercontent.com/761780/80909569-e1ea7200-8d5b-11ea-8a1a-5dbaf87e236b.png)

**Comment by hollasch** [comment]
Refraction bugs are typically fun to figure out. One thing to try is to hand trace a ray directly through the center of the glass sphere.

**Comment by maurocolella** [comment]
Thank you. I was concerned that there could be a bug or room for clarification in the relevant chapter of the book. I will take a close look as soon as I have time.
